### PR TITLE
Fixes #26050 - yum exclude list

### PR DESCRIPTION
**Pull request by upadhyeammit**
This adds check for yum exclude list, if its configured then user gets error with message. The message includes which file has configured it and what is value of same.

### CODE REVIEW

**Comment by mbacovsky** [comment]
@upadhyeammit is the right issue linked to the PR. Shouldn't it be https://projects.theforeman.org/issues/26050?

**Comment by upadhyeammit** [comment]
@mbacovsky my mistake :( changed it ! Thank You !

**Comment by jameerpathan111** [comment]
@upadhyeammit I have tested this pr and found below issue.
- I have added `exclude=kernel* abc*` in `/etc/yum.conf`,
the check fails as expected but the issue is that it doesn't show complete list.
I think we should change `grep -w exclude` in such a way that we can get complete list.

```
--------------------------------------------------------------------------------
Check if yum exclude list is configured:                              [FAIL]
The /etc/yum.conf has exclude list configured as below,
  exclude = kernel*,
Unset this as it can cause yum update or upgrade failures !
--------------------------------------------------------------------------------
```

**Comment by upadhyeammit** [comment]
@jameerpathan111 I switched to yum.conf which is fixing your concern. Request to check.
Thank You !

**Comment by upadhyeammit** [comment]
@jameerpathan111 I can see one more issue with this pr, if yum.conf not having exclude defined it will fail ! The yum-config-manager use to help with this situation, I am checking this one further.

**Comment by iNecas** [comment]
What is the status here?

**Comment by upadhyeammit** [comment]
> What is the status here?

From my side work is complete, when we had last meeting @mbacovsky mentioned that this may need changes because of version lock PR .

**Comment by iNecas** [comment]
@mbacovsky mind having a look?

File: definitions/checks/yum_exclude.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  def run
+    assert(exclude_set?, 'The /etc/yum.conf has exclude list configured as below,'\
+          "\n  #{yum_exclude}"\
+          "\nUnset this as it can cause yum update or upgrade failures !")
+  end
+
+  def exclude_set?
+    yum_exclude == 'exclude=' || yum_exclude == 'exclude ='
```
Comment (mbacovsky) [inline definitions/checks/yum_exclude.rb]: This call the grep twice in many cases and misses e.g. '# exclude=kernel*'. Could we use regexp instead like
```ruby
yum_exclude =~ /^exclude\s*=\s*\S+/
```

File: definitions/checks/yum_exclude.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  def yum_exclude
+    execute!('grep -w exclude /etc/yum.conf')
+  rescue ForemanMaintain::Error::ExecutionError
+    'exclude='
```
Comment (mbacovsky) [inline definitions/checks/yum_exclude.rb]: This feels quite hacky, could we return empty string here?
Perhaps we could call the method `grep_yum_exclude` to see better what it does.

**Comment by upadhyeammit** [comment]
> @upadhyeammit I did some more testing and it behaves almost as expected. However I took me quite some time to understand why. I've suggested some changes inside.
> 
> As for our discussion on using the package manager layer - I don't think we need to do that now. The functionality is nicely bundled in one file and I have no idea how excluding of packages works in Debian world so I guess it would slip this for too long. What do you think?

I am also not familiar with how Debian handles excluding packages, I would explore on it but now what I feel is we should merge this for yum after taking care of suggestions added by you. I will check for Debian side and would be including it next.

File: definitions/checks/yum_exclude.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+class Checks::YumExclude < ForemanMaintain::Check
+  metadata do
+    label :check_yum_exclude_list
+    description 'Check if yum exclude list is configured'
+    tags :pre_upgrade
+  end
+
+  def run
+    assert(exclude_set?, 'The /etc/yum.conf has exclude list configured as below,'\
```
[the remark this answers is not in the document]
Comment (upadhyeammit) [reply definitions/checks/yum_exclude.rb]: I changed it to return true when exclude is set, sorry for confusion !

File: definitions/checks/yum_exclude.rb (hunk 1 above)
Comment (upadhyeammit) [reply definitions/checks/yum_exclude.rb]: I changed this to regular expression, miss from my side about commented exclude line.

File: definitions/checks/yum_exclude.rb (hunk 2 above)
Comment (upadhyeammit) [reply definitions/checks/yum_exclude.rb]: I changed the method name and now using execute_with_status which returns empty string when exclude is not set.

**Comment by upadhyeammit** [comment]
@mbacovsky with recent commit only thing is we call grep twice when exclude is set, I was trying to avoid it but it was unnecessarily adding more lines to check and I feel one extra grep command call should not be that expensive. Request to review once ! thank you !

**Comment by mbacovsky** [comment]
@upadhyeammit thank you for the update and incorporating my comments. I checked on the PR and it looks good. However I'm still not very convinced about the unnecessary double grep call in the check (the execution is logged twice, new subshell is executed etc) so I'm sorry but I'd appreciate if we can get rid of it. Could you please reconsider it?

**Comment by upadhyeammit** [comment]
> @upadhyeammit thank you for the update and incorporating my comments. I checked on the PR and it looks good. However I'm still not very convinced about the unnecessary double grep call in the check (the execution is logged twice, new subshell is executed etc) so I'm sorry but I'd appreciate if we can get rid of it. Could you please reconsider it?

No problem at all, will send fix for same ! thank you for review

**Comment by upadhyeammit** [comment]
@mbacovsky did changes as requested, thank you for help ! Request to have a look

**Comment by mbacovsky** [comment]
@upadhyeammit thank you for the update, I really appreciate it. I'll merge this once the tests finish and are green.

**Comment by upadhyeammit** [comment]
> @upadhyeammit thank you for the update, I really appreciate it. I'll merge this once the tests finish and are green.

ack! Travis was failing due to freeze for regex, I removed it and it should be green now!

**Comment by mbacovsky** [comment]
Thanks again @upadhyeammit!

**Pull request merged**